sinks/hawkular/driver.go

```diff
@@ -0,0 +1,236 @@
+		if len(l.Description) > 0 {
+			tags[l.Key+desc_tag] = l.Description
+		}
+	}
+
+	if len(md.Units.String()) > 0 {
+		tags[units_tag] = md.Units.String()
+	}
+	if len(md.Type.String()) > 0 {
+		tags[type_tag] = md.Type.String()
+	}
+
+	hmd := metrics.MetricDefinition{
+		Id:   md.Name,
+		Tags: tags,
+	}
+
+	return hmd
+}
+
+func (self *hawkularSink) groupName(p *sink_api.Point) string {
+	n := []string{p.Labels[sink_api.LabelContainerName], p.Name}
+	return strings.Join(n, sep_str)
+}
+
+func (self *hawkularSink) idName(p *sink_api.Point) string {
+	n := []string{p.Labels[sink_api.LabelContainerName], p.Labels[sink_api.LabelPodId], p.Name}
+	return strings.Join(n, sep_str)
+}
+
+// Check that metrics tags are defined on the Hawkular server and if not,
+// register the metric definition.
+func (self *hawkularSink) registerIfNecessary(t *sink_api.Timeseries) error {
+	key := self.idName(t.Point)
+
+	self.regLock.Lock()
+	defer self.regLock.Unlock()
+
+	// If found, check it matches the current stored definition (could be old info from
+	// the stored metrics cache for example)
+	if _, found := self.reg[key]; !found {
+		// Register the metric descriptor here..
+		if md, f := self.models[t.MetricDescriptor.Name]; f {
+			// Set tag values
+			for k, v := range t.Point.Labels {
+				md.Tags[k] = v
+			}
+
+			md.Tags[group_tag] = self.groupName(t.Point)
+
+			// Create metric
+			if err := self.client.UpdateTags(metrics.Gauge, t.MetricDescriptor.Name, md.Tags); err != nil {
+				// Log error and don't add this key to the lookup table
+				return err
+			}
+
+			// Add to the lookup table
+			self.reg[key] = &md
+			glog.Infof("Registered new metric definition: %s", key)
+		} else {
+			return fmt.Errorf("Could not find definition model with name %s", t.MetricDescriptor.Name)
+		}
+	}
+	// TODO Compare the definition tags and update if necessary? Quite expensive operation..
+
+	return nil
+}
+
+func (self *hawkularSink) StoreTimeseries(ts []sink_api.Timeseries) error {
+	mhs := make([]metrics.MetricHeader, 0, len(ts))
+
+	for _, t := range ts {
+		self.registerIfNecessary(&t)
+
+		if t.MetricDescriptor.ValueType == sink_api.ValueBool {
+			// TODO: Model to availability type once we see some real world examples
+			break
+		}
+
+		mH, err := self.pointToMetricHeader(&t)
+		if err != nil {
+			// One transformation error should not prevent the whole process
+			glog.Errorf(err.Error())
+			continue
+		}
+
+		mhs = append(mhs, *mH)
+	}
+
+	return self.client.Write(mhs)
+}
+
+// Converts Timeseries to metric structure used by the Hawkular
+func (self *hawkularSink) pointToMetricHeader(t *sink_api.Timeseries) (*metrics.MetricHeader, error) {
+
+	p := t.Point
+	name := self.idName(p)
+
+	value, err := metrics.ConvertToFloat64(p.Value)
+	if err != nil {
+		return nil, err
+	}
+
+	m := metrics.Datapoint{
+		Value:     value,
+		Timestamp: metrics.UnixMilli(p.End),
+	}
+
+	// At the moment all the values are converted to gauges.
```

Comment:
Our gauge support fits better these currently collected metrics than our counter model would. This might or might not change in the future, but for now that's the case. Also, there weren't any Boolean values yet so modeling them would be quite premature.